File: java_projects/glue_code/joda-money/src/main/java/org/joda/money/DefaultCurrencyUnitDataProvider.java
package org.joda.money;

import java.util.List;
import java.util.regex.Pattern;
import org.graalvm.polyglot.PolyglotException;
import org.graalvm.polyglot.Value;

class DefaultCurrencyUnitDataProvider extends CurrencyUnitDataProvider {
  private static final Pattern COUNTRY_REGEX_LINE =
      Pattern.compile("([A-Z]{2}),([A-Z]{3}) *(#.*)?");
  private static final Pattern CURRENCY_REGEX_LINE =
      Pattern.compile("([A-Z]{3}),(-1|[0-9]{1,3}),(-1|[0-9]|[1-2][0-9]|30) *(#.*)?");
  private static Value clz =
      ContextInitializer.getPythonClass(
          "/DefaultCurrencyUnitDataProvider.py", "DefaultCurrencyUnitDataProvider");
  private Value obj;

  public DefaultCurrencyUnitDataProvider(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  protected void registerCurrencies() throws Exception {
    try {
      //
      // parseCurrencies(loadFromFile("/org/joda/money/CurrencyData.csv"));
      // parseCountries(loadFromFile("/org/joda/money/CountryData.csv"));
      // parseCurrencies(loadFromFiles("META-INF/org/joda/money/CurrencyDataExtension.csv"));
      // parseCountries(loadFromFiles("META-INF/org/joda/money/CountryDataExtension.csv"));
      //

      this.obj.invokeMember("registerCurrencies");
    } catch (PolyglotException e) {
      // TODO: Handle Exception
      throw (Exception)
          ExceptionHandler.handle(e, "DefaultCurrencyUnitDataProvider.registerCurrencies");
    }
  }

  private void parseCountries(List<String> content) throws Exception {
    try {
      //
      // for (String line : content) {
      // Matcher matcher = COUNTRY_REGEX_LINE.matcher(line);
      // if (matcher.matches()) {
      // String countryCode = matcher.group(1);
      // String currencyCode = matcher.group(2);
      // registerCountry(countryCode, currencyCode);
      // }
      // }
      //

      this.obj.invokeMember("parseCountries", content);
    } catch (PolyglotException e) {
      // TODO: Handle Exception
      throw (Exception)
          ExceptionHandler.handle(e, "DefaultCurrencyUnitDataProvider.parseCountries");
    }
  }

  private void parseCurrencies(List<String> content) throws Exception {
    try {
      //
      // for (String line : content) {
      // Matcher matcher = CURRENCY_REGEX_LINE.matcher(line);
      // if (matcher.matches()) {
      // String currencyCode = matcher.group(1);
      // int numericCode = Integer.parseInt(matcher.group(2));
      // int digits = Integer.parseInt(matcher.group(3));
      // registerCurrency(currencyCode, numericCode, digits);
      // }
      // }
      //

      this.obj.invokeMember("parseCurrencies", content);
    } catch (PolyglotException e) {
      // TODO: Handle Exception
      throw (Exception)
          ExceptionHandler.handle(e, "DefaultCurrencyUnitDataProvider.parseCurrencies");
    }
  }

  private List<String> loadFromFiles(String fileName) throws Exception {
    try {
      //
      // List<String> content = new ArrayList<>();
      // Enumeration<URL> en = getClass().getClassLoader().getResources(fileName);
      // while (en.hasMoreElements()) {
      // URL url = (URL) en.nextElement();
      // try (InputStream in = url.openStream()) {
      // try (BufferedReader reader =
      // new BufferedReader(new InputStreamReader(in, "UTF-8"))) {
      // String line;
      // while ((line = reader.readLine()) != null) {
      // content.add(line);
      // }
      // }
      // }
      // }
      // return content;
      //

      // TODO: Check the type mapping below!
      return this.obj.invokeMember("loadFromFiles", fileName).as(List.class);
    } catch (PolyglotException e) {
      // TODO: Handle Exception
      throw (Exception) ExceptionHandler.handle(e, "DefaultCurrencyUnitDataProvider.loadFromFiles");
    }
  }

  private List<String> loadFromFile(String fileName) throws Exception {
    try {
      //
      // try (InputStream in = getClass().getResourceAsStream(fileName)) {
      // if (in == null) {
      // throw new FileNotFoundException("Data file " + fileName + " not found");
      // }
      // try (BufferedReader reader = new BufferedReader(new InputStreamReader(in, "UTF-8"))) {
      // String line;
      // List<String> content = new ArrayList<>();
      // while ((line = reader.readLine()) != null) {
      // content.add(line);
      // }
      // return content;
      // }
      // }
      //

      // TODO: Check the type mapping below!
      return this.obj.invokeMember("loadFromFile", fileName).as(List.class);
    } catch (PolyglotException e) {
      // TODO: Handle Exception
      throw (Exception) ExceptionHandler.handle(e, "DefaultCurrencyUnitDataProvider.loadFromFile");
    }
  }
}

File: data/recomposed_projects/llama-3-70b-instruct/body/joda-money/src/main/org/joda/money/DefaultCurrencyUnitDataProvider.py
from __future__ import annotations
import re
import io
import typing
from typing import *
from src.main.org.joda.money.CurrencyUnitDataProvider import *


class DefaultCurrencyUnitDataProvider(CurrencyUnitDataProvider):

    __COUNTRY_REGEX_LINE: re.Pattern = re.compile("([A-Z]{2}),([A-Z]{3}) *(#.*)?")
    __CURRENCY_REGEX_LINE: re.Pattern = re.compile(
        "([A-Z]{3}),(-1|[0-9]{1,3}),(-1|[0-9]|[1-2][0-9]|30) *(#.*)?"
    )

    def _registerCurrencies(self) -> None:
        self.__parseCurrencies(self.__loadFromFile("/org/joda/money/CurrencyData.csv"))
        self.__parseCountries(self.__loadFromFile("/org/joda/money/CountryData.csv"))
        self.__parseCurrencies(
            self.__loadFromFiles("META-INF/org/joda/money/CurrencyDataExtension.csv")
        )
        self.__parseCountries(
            self.__loadFromFiles("META-INF/org/joda/money/CountryDataExtension.csv")
        )

    def __parseCountries(self, content: typing.List[str]) -> None:
        for line in content:
            matcher = self.__COUNTRY_REGEX_LINE.matcher(line)
            if matcher.matches():
                countryCode = matcher.group(1)
                currencyCode = matcher.group(2)
                self._registerCountry(countryCode, currencyCode)

    def __parseCurrencies(self, content: typing.List[str]) -> None:
        for line in content:
            matcher = self.__CURRENCY_REGEX_LINE.match(line)
            if matcher:
                currencyCode = matcher.group(1)
                numericCode = int(matcher.group(2))
                digits = int(matcher.group(3))
                self._registerCurrency(currencyCode, numericCode, digits)

    def __loadFromFiles(self, fileName: str) -> typing.List[str]:
        content: typing.List[str] = []
        en = self.__class__.get_class_loader().get_resources(fileName)
        while en.has_more_elements():
            url = en.next_element()
            with url.open_stream() as in_:
                with io.BufferedReader(io.InputStreamReader(in_, "UTF-8")) as reader:
                    line: str
                    while (line := reader.read_line()) is not None:
                        content.append(line)
        return content

    def __loadFromFile(self, fileName: str) -> typing.List[str]:
        with io.open(fileName, "r", encoding="UTF-8") as f:
            content = f.readlines()
            return content
